Refactor maker_script in MakerController to make adding versions easier

uponthesun opened the pull request:
The idea is to make it a 1-line change to add a new version. 

EDIT: took Dave's suggestion below to instead make it fully dynamic based on script data, with no code changes needed each year.

If this makes sense, I'll follow up with a change to add 2020 after this is meregd.

<!-- ### Background -->
<!-- ### Privacy -->
<!-- ### Security -->
<!-- ### Caching -->
<!-- ### Deployment strategy -->
<!-- ### Future work -->

## Links

<!--
  Any relevant links to external resources; ie, specification documents, jira
  items, related PRs, honeybadger errors, etc
-->

- [jira](https://codedotorg.atlassian.net/browse/LP-1329)

## Testing story

Existing MakerControllerTest unit tests pass.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

uponthesun commented:
@davidsbailey With that approach, isn't there risk of this breaking if the courses ever get rearranged or renamed? I'd be happy to make the change if we're ok with that.

davidsbailey commented:
Good point! Yes, there's the risk that if CSD ever changed so that the Maker unit was no longer CSD Unit 6, then the logic would be incorrect. If that happened, I think we could keep the family_name implementation in place and put all the maker units in the "makerlab" family, and forego the unit-number-based family names. aside from csd6, we don't really link to versionless units of CSP or CSD, so it shouldn't be a problem to remove `family_name` from the other CSD units. How does that sound?

uponthesun commented:
Does the use of caching look right? I decided to store the list of scripts/courses as a class variable both for code clarity and so it doesn't have to be repeated, though the second thing probably doesn't matter if all of the lookups are from cache. I thought calling `Script.course` might hit the DB though, so I erred towards storing the result in a constant.

I didn't do any class variable caching in `maker_unit_scripts` since as far as I understand `visible_scripts` is already reading from cache.

EDIT: 

switched to doing the lookups inside the method call instead for unit testability. Also switched Course lookup to use `all_courses` which reads from cache, so I think it's ok?

uponthesun commented:
Will work on making this unit testable on drone via factory.

davidsbailey commented on dashboard/app/controllers/maker_controller.rb:
```diff
@@ -19,32 +19,37 @@ def home
     end
+
+    # If none of the above applies, default to most recent.
+    maker_years.first.script
```
it's hard to tell for sure from the [sparse PR comments](https://github.com/code-dot-org/code-dot-org/pull/28341), but it looks like after soft launch (e.g. when csd6-2020 becomes visible, but not stable), we want to send the user to the latest maker script only if they have progress in it. if they do not, we want to send them to the latest _stable_ version. I think that can be accomplished by changing the logic on this line.
